Complete unqualified enum names in patterns and expressions

nathanwhit opened the pull request:
This PR implements the completion described in #4014.
The result looks like so for patterns:
<img width="542" alt="Screen Shot 2020-04-20 at 3 53 55 PM" src="https://user-images.githubusercontent.com/17734409/79794010-8f529400-831f-11ea-9673-f838aa9bc962.png">

and for `expr`s:
<img width="620" alt="Screen Shot 2020-04-21 at 3 51 24 PM" src="https://user-images.githubusercontent.com/17734409/79908784-d73ded80-83e9-11ea-991d-921f0cb27e6f.png">


I'm not confident that the completion text itself is very robust, as it will unconditionally add completions for enum variants with the form `Enum::Variant`. This means (I believe) it would still suggest `Enum::Variant` even if the local name is changed i.e. `use Enum as Foo` or the variants are brought into scope such as through `use Enum::*`.

flodiebold commented on crates/ra_ide/src/completion/complete_unqualified_path.rs:
```diff
@@ -31,6 +56,57 @@ pub(super) fn complete_unqualified_path(acc: &mut Completions, ctx: &CompletionC
+            if let Some(Adt::Enum(enum_data)) = ty.as_adt() {
+                let variants = enum_data.variants(ctx.db);
+                for variant in variants {
+                    let mut local_name = enum_data.name(ctx.db).to_string();
```
we can use `find_use_path` to instead build a proper path to the enum that will take reexports etc. into account.

flodiebold commented on crates/ra_ide/src/completion/complete_unqualified_path.rs:
```diff
@@ -4,17 +4,42 @@ use hir::ScopeDef;
-        || ctx.record_pat_syntax.is_some()
-    {
+    for ancestor in ctx.token.ancestors() {
+        if let Some(()) = match ancestor.kind() {
```
I don't understand why this should be necessary -- why not just look for a `Pat` in the ancestors?

Also, I'd suggest moving this computation of the 'expected type' to `CompletionContext`.

nathanwhit replied:
Not sure why I didn't think to try that, it seems obvious in retrospect! That also simplifies the code greatly.

flodiebold commented on crates/ra_ide/src/completion/complete_unqualified_path.rs:
```diff
@@ -4,17 +4,42 @@ use hir::ScopeDef;
+use ra_syntax::{ast, AstNode, SyntaxKind};
 
 pub(super) fn complete_unqualified_path(acc: &mut Completions, ctx: &CompletionContext) {
-    if !ctx.is_trivial_path {
```
not sure about removing this condition -- if we already have a qualified path like `Enum::Qu<|>`, there's other completions that will complete the enum variant

flodiebold replied:
Oh, and we don't need to worry about `use Enum::*`; in that case the normal unqualified path completion will already do the job :)

nathanwhit replied:
Regarding the `is_trivial_path` condition, I found that whenever the cursor is in a pattern position the `is_trivial_path` is set to `false`. I think this is because [here](https://github.com/rust-analyzer/rust-analyzer/blob/7ab28cacbb50d78e4e6f12293351683625cab607/crates/ra_ide/src/completion/completion_context.rs#L296-L297) the `CompletionContext` only checks the path and sets that flag if the current node is a `NameRef`, and it doesn't do so for `BindPat`s.

For this PR I changed the condition to `return` early only if it is not a trivial path _and_ not a pattern.

flodiebold replied:
> I found that whenever the cursor is in a pattern position the is_trivial_path is set to false.

Hmm, that might be a bug -- I vaguely recall frequently being annoyed about not getting completions I'd expect in patterns, though I haven't looked into it :thinking:

nathanwhit commented:
Updated the PR to apply the given feedback. Due to the changes it made completion for enum variants in `expr`s trivial, so that now works as well! I've updated the PR message to reflect this, as well as added a screenshot of what that looks like.

matklad commented on crates/ra_ide/src/completion/complete_unqualified_path.rs:
```diff
@@ -4,20 +4,28 @@ use hir::ScopeDef;
         return;
     }
 
+    for ancestor in ctx.token.ancestors() {
```
Hm, why do we need a loop here? I'd probably expect this to be just an `if let Some(ty) = ctx.expected_type_of(&node) {`, without the loop. Though, we might need a loop inside `expected_type_of`.

nathanwhit replied:
Good point! I've moved the loop into `expected_type_of` now.

matklad commented:
bors r+

I am not sure that conditions and `expected_type` are 100% right, I think we might want to tweak things there. OTOH, it seems good enough to merge now
